File: nix/packages/rice/test/cli.test.ts
import assert from "node:assert/strict";

import { test } from "bun:test";

import { parseCli, parseCliArgv, renderTopLevelHelp } from "../src/core/cli.ts";

const expectHelp = (argv: readonly string[], needle: RegExp): void => {
  const parsed = parseCliArgv(argv);
  assert.equal(parsed._tag, "Help");
  assert.equal(parsed.exitCode, 0);
  assert.match(parsed.text, needle);
};

const expectError = (
  argv: readonly string[],
  message: string,
  usageNeedle: RegExp,
): void => {
  const parsed = parseCliArgv(argv);
  assert.equal(parsed._tag, "Error");
  assert.equal(parsed.exitCode, 1);
  assert.equal(parsed.message, message);
  assert.match(parsed.text ?? "", usageNeedle);
};

test("parse top-level help when no args are provided", () => {
  const parsed = parseCliArgv(["rice"]);
  assert.equal(parsed._tag, "Help");
  assert.equal(parsed.text, renderTopLevelHelp());
});

test("parse top-level help for explicit help flag", () => {
  expectHelp(["rice", "--help"], /Usage: rice <COMMAND>/);
  expectHelp(["--help"], /Usage: rice <COMMAND>/);
});

test("parse accepts argv without the binary prefix", () => {
  const cli = parseCli(["system", "build"]);
  assert.deepEqual(cli, {
    command: {
      _tag: "System",
      command: "Build",
    },
  });
});

test("parse home build keeps default host when only user is provided", () => {
  const cli = parseCli(["rice", "home", "build", "alice"]);
  assert.equal(cli.command._tag, "Home");
  assert.equal(cli.command.command._tag, "Build");
  assert.equal(cli.command.command.user, "alice");
  assert.equal(cli.command.command.host, "wsl");
});

test("parse home switch uses defaults when user and host are omitted", () => {
  const cli = parseCli(["rice", "home", "switch"]);
  assert.equal(cli.command._tag, "Home");
  assert.equal(cli.command.command._tag, "Switch");
  assert.equal(cli.command.command.user, "annt");
  assert.equal(cli.command.command.host, "wsl");
});

test("parse home command treats trailing help flag as group help", () => {
  expectHelp(["rice", "home", "build", "--help"], /Usage: rice home <COMMAND> \[USER\] \[HOST\]/);
});

for (const [name, argv, usage] of [
  ["system group help on missing subcommand", ["rice", "system"], /Usage: rice system <COMMAND>/],
  ["home group help on help flag", ["rice", "home", "--help"], /Usage: rice home <COMMAND> \[USER\] \[HOST\]/],
  ["nixos group help on missing subcommand", ["rice", "nixos"], /Usage: rice nixos <COMMAND>/],
  ["darwin group help on help flag", ["rice", "darwin", "--help"], /Usage: rice darwin <COMMAND>/],
  ["nix group help on missing subcommand", ["rice", "nix"], /Usage: rice nix <COMMAND>/],
  ["flake group help on help flag", ["rice", "flake", "--help"], /Usage: rice flake <COMMAND>/],
] as const) {
  test(name, () => {
    expectHelp(argv, usage);
  });
}

for (const [name, argv, message, usage] of [
  ["unknown top-level command errors", ["rice", "wat"], "unknown command: wat", /Usage: rice <COMMAND>/],
  ["unknown system subcommand errors", ["rice", "system", "wat"], "unknown system subcommand: wat", /Usage: rice system <COMMAND>/],
  ["unknown home subcommand errors", ["rice", "home", "wat"], "unknown home subcommand: wat", /Usage: rice home <COMMAND> \[USER\] \[HOST\]/],
  ["unknown nixos subcommand errors", ["rice", "nixos", "wat"], "unknown nixos subcommand: wat", /Usage: rice nixos <COMMAND>/],
  ["unknown darwin subcommand errors", ["rice", "darwin", "wat"], "unknown darwin subcommand: wat", /Usage: rice darwin <COMMAND>/],
  ["unknown nix subcommand errors", ["rice", "nix", "wat"], "unknown nix subcommand: wat", /Usage: rice nix <COMMAND>/],
  ["unknown flake subcommand errors", ["rice", "flake", "wat"], "unknown flake subcommand: wat", /Usage: rice flake <COMMAND>/],
] as const) {
  test(name, () => {
    expectError(argv, message, usage);
  });
}

for (const [name, argv, message, usage] of [
  ["system build rejects extra args", ["rice", "system", "build", "extra"], "unexpected arguments: extra", /Usage: rice system <COMMAND>/],
  ["home build rejects too many args", ["rice", "home", "build", "a", "b", "c"], "unexpected arguments: c", /Usage: rice home <COMMAND> \[USER\] \[HOST\]/],
  ["nixos build rejects extra args", ["rice", "nixos", "build", "extra"], "unexpected arguments: extra", /Usage: rice nixos <COMMAND>/],
  ["darwin switch rejects extra args", ["rice", "darwin", "switch", "extra"], "unexpected arguments: extra", /Usage: rice darwin <COMMAND>/],
  ["nix clean rejects extra args", ["rice", "nix", "clean", "extra"], "unexpected arguments: extra", /Usage: rice nix <COMMAND>/],
  ["flake check rejects extra args", ["rice", "flake", "check", "extra"], "unexpected arguments: extra", /Usage: rice flake <COMMAND>/],
  ["flake update rejects extra args", ["rice", "flake", "update", "fenix", "extra"], "unexpected arguments: extra", /Usage: rice flake <COMMAND>/],
] as const) {
  test(name, () => {
    const parsed = parseCliArgv(argv);
    assert.equal(parsed._tag, "Error");
    assert.equal(parsed.message, message);
    assert.match(parsed.text ?? "", usage);
  });
}

test("flake update requires an input name", () => {
  expectError(
    ["rice", "flake", "update"],
    'missing flake input name for "update"',
    /Usage: rice flake <COMMAND>/,
  );
});

test("flake update treats help flag as a missing input name", () => {
  expectError(
    ["rice", "flake", "update", "--help"],
    'missing flake input name for "update"',
    /Usage: rice flake <COMMAND>/,
  );
});

test("parseCli throws on help result", () => {
  assert.throws(() => parseCli(["rice"]), new Error("expected parsed CLI"));
});

test("parseCli throws with the parser error message", () => {
  assert.throws(() => parseCli(["rice", "wat"]), new Error("unknown command: wat"));
});

File: nix/packages/rice/internal/rice/cli_test.go
package rice

import (
	"strings"
	"testing"
)

func TestParse(t *testing.T) {
	cases := []struct {
		name     string
		args     []string
		wantKind string
		// success
		wantTag     string
		wantSubcmd  string
		wantUser    string
		wantHost    string
		wantHasHost bool
		wantName    string
		// help/error
		wantTextContains []string
		wantMsgExact     string
		wantMsgContains  string
	}{
		// ── Top-level help ──
		{
			name:             "no args",
			args:             nil,
			wantKind:         "help",
			wantTextContains: []string{"Usage: rice <COMMAND>"},
		},
		{
			name:     "rice --help",
			args:     []string{"rice", "--help"},
			wantKind: "help",
		},
		{
			name:     "--help",
			args:     []string{"--help"},
			wantKind: "help",
		},
		{
			name:     "-h",
			args:     []string{"-h"},
			wantKind: "help",
		},

		// ── Successful command parsing ──
		{
			name:       "system build",
			args:       []string{"rice", "system", "build"},
			wantKind:   "success",
			wantTag:    "system",
			wantSubcmd: "build",
		},
		{
			name:       "system switch",
			args:       []string{"rice", "system", "switch"},
			wantKind:   "success",
			wantTag:    "system",
			wantSubcmd: "switch",
		},
		{
			name:       "nixos boot",
			args:       []string{"rice", "nixos", "boot"},
			wantKind:   "success",
			wantTag:    "nixos",
			wantSubcmd: "boot",
		},
		{
			name:       "darwin build",
			args:       []string{"rice", "darwin", "build"},
			wantKind:   "success",
			wantTag:    "darwin",
			wantSubcmd: "build",
		},
		{
			name:       "nix optimise",
			args:       []string{"rice", "nix", "optimise"},
			wantKind:   "success",
			wantTag:    "nix",
			wantSubcmd: "optimise",
		},
		{
			name:       "nix repair",
			args:       []string{"rice", "nix", "repair"},
			wantKind:   "success",
			wantTag:    "nix",
			wantSubcmd: "repair",
		},
		{
			name:       "nix clean",
			args:       []string{"rice", "nix", "clean"},
			wantKind:   "success",
			wantTag:    "nix",
			wantSubcmd: "clean",
		},
		{
			name:       "flake check",
			args:       []string{"rice", "flake", "check"},
			wantKind:   "success",
			wantTag:    "flake",
			wantSubcmd: "check",
		},
		{
			name:       "flake fmt",
			args:       []string{"rice", "flake", "fmt"},
			wantKind:   "success",
			wantTag:    "flake",
			wantSubcmd: "fmt",
		},
		{
			name:       "flake update all",
			args:       []string{"rice", "flake", "update", "all"},
			wantKind:   "success",
			wantTag:    "flake",
			wantSubcmd: "update",
			wantName:   "all",
		},
		{
			name:       "flake update nixpkgs",
			args:       []string{"rice", "flake", "update", "nixpkgs"},
			wantKind:   "success",
			wantTag:    "flake",
			wantSubcmd: "update",
			wantName:   "nixpkgs",
		},
		{
			name:        "home build default",
			args:        []string{"rice", "home", "build"},
			wantKind:    "success",
			wantTag:     "home",
			wantSubcmd:  "build",
			wantUser:    "annt",
			wantHasHost: false,
		},
		{
			name:        "home build user",
			args:        []string{"rice", "home", "build", "alice"},
			wantKind:    "success",
			wantTag:     "home",
			wantSubcmd:  "build",
			wantUser:    "alice",
			wantHasHost: false,
		},
		{
			name:        "home build user host",
			args:        []string{"rice", "home", "build", "alice", "mbp"},
			wantKind:    "success",
			wantTag:     "home",
			wantSubcmd:  "build",
			wantUser:    "alice",
			wantHost:    "mbp",
			wantHasHost: true,
		},
		{
			name:        "home switch default",
			args:        []string{"rice", "home", "switch"},
			wantKind:    "success",
			wantTag:     "home",
			wantSubcmd:  "switch",
			wantUser:    "annt",
			wantHasHost: false,
		},
		{
			name:        "home switch user host",
			args:        []string{"rice", "home", "switch", "bob", "nixbox"},
			wantKind:    "success",
			wantTag:     "home",
			wantSubcmd:  "switch",
			wantUser:    "bob",
			wantHost:    "nixbox",
			wantHasHost: true,
		},

		// ── Group help ──
		{
			name:             "system group help",
			args:             []string{"rice", "system"},
			wantKind:         "help",
			wantTextContains: []string{"Usage: rice system"},
		},
		{
			name:     "system --help",
			args:     []string{"rice", "system", "--help"},
			wantKind: "help",
		},
		{
			name:     "nixos group help",
			args:     []string{"rice", "nixos"},
			wantKind: "help",
		},
		{
			name:     "nixos -h",
			args:     []string{"rice", "nixos", "-h"},
			wantKind: "help",
		},
		{
			name:     "darwin group help",
			args:     []string{"rice", "darwin"},
			wantKind: "help",
		},
		{
			name:     "nix group help",
			args:     []string{"rice", "nix"},
			wantKind: "help",
		},
		{
			name:     "home group help",
			args:     []string{"rice", "home"},
			wantKind: "help",
		},
		{
			name:     "flake group help",
			args:     []string{"rice", "flake"},
			wantKind: "help",
		},
		{
			name:     "flake --help",
			args:     []string{"rice", "flake", "--help"},
			wantKind: "help",
		},

		// ── Special help cases (preserve TS oddities) ──
		{
			name:     "home build --help",
			args:     []string{"rice", "home", "build", "--help"},
			wantKind: "help",
		},
		{
			name:            "flake update --help",
			args:            []string{"rice", "flake", "update", "--help"},
			wantKind:        "error",
			wantMsgContains: "missing flake input name",
		},

		// ── Error cases ──
		{
			name:             "unknown top-level",
			args:             []string{"rice", "wat"},
			wantKind:         "error",
			wantMsgExact:     "unknown command: wat",
			wantTextContains: []string{"Usage: rice <COMMAND>"},
		},
		{
			name:         "unknown system subcmd",
			args:         []string{"rice", "system", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown system subcommand: wat",
		},
		{
			name:         "unknown home subcmd",
			args:         []string{"rice", "home", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown home subcommand: wat",
		},
		{
			name:         "unknown nixos subcmd",
			args:         []string{"rice", "nixos", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown nixos subcommand: wat",
		},
		{
			name:         "unknown darwin subcmd",
			args:         []string{"rice", "darwin", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown darwin subcommand: wat",
		},
		{
			name:         "unknown nix subcmd",
			args:         []string{"rice", "nix", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown nix subcommand: wat",
		},
		{
			name:         "unknown flake subcmd",
			args:         []string{"rice", "flake", "wat"},
			wantKind:     "error",
			wantMsgExact: "unknown flake subcommand: wat",
		},
		{
			name:         "system build extra",
			args:         []string{"rice", "system", "build", "extra"},
			wantKind:     "error",
			wantMsgExact: "unexpected arguments: extra",
		},
		{
			name:     "nixos switch extra",
			args:     []string{"rice", "nixos", "switch", "x", "y"},
			wantKind: "error",
		},
		{
			name:     "home build three extra",
			args:     []string{"rice", "home", "build", "a", "b", "c"},
			wantKind: "error",
		},
		{
			name:            "flake update missing",
			args:            []string{"rice", "flake", "update"},
			wantKind:        "error",
			wantMsgContains: "missing flake input name",
		},

		// ── Leading-hyphen name hardening ──
		{
			name:            "flake update -bad",
			args:            []string{"rice", "flake", "update", "-bad"},
			wantKind:        "error",
			wantMsgContains: "invalid flake input name",
		},
		{
			name:            "flake update empty",
			args:            []string{"rice", "flake", "update", ""},
			wantKind:        "error",
			wantMsgContains: "invalid flake input name",
		},

		// ── Accepts argv without rice prefix ──
		{
			name:       "system build no prefix",
			args:       []string{"system", "build"},
			wantKind:   "success",
			wantTag:    "system",
			wantSubcmd: "build",
		},
		{
			name:       "home build no prefix",
			args:       []string{"home", "build"},
			wantKind:   "success",
			wantTag:    "home",
			wantSubcmd: "build",
		},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			got := Parse(tt.args)

			if got.Kind != tt.wantKind {
				t.Errorf("Kind = %q, want %q", got.Kind, tt.wantKind)
			}

			if tt.wantKind == "success" {
				if got.CLI.Tag != tt.wantTag {
					t.Errorf("CLI.Tag = %q, want %q", got.CLI.Tag, tt.wantTag)
				}
				if got.CLI.Subcmd != tt.wantSubcmd {
					t.Errorf("CLI.Subcmd = %q, want %q", got.CLI.Subcmd, tt.wantSubcmd)
				}
				if tt.wantUser != "" || tt.wantHasHost {
					if got.CLI.User != tt.wantUser {
						t.Errorf("CLI.User = %q, want %q", got.CLI.User, tt.wantUser)
					}
				}
				if tt.wantHost != "" || tt.wantHasHost {
					if got.CLI.Host != tt.wantHost {
						t.Errorf("CLI.Host = %q, want %q", got.CLI.Host, tt.wantHost)
					}
					if got.CLI.HasHost != tt.wantHasHost {
						t.Errorf("CLI.HasHost = %v, want %v", got.CLI.HasHost, tt.wantHasHost)
					}
				}
				if tt.wantName != "" {
					if got.CLI.Name != tt.wantName {
						t.Errorf("CLI.Name = %q, want %q", got.CLI.Name, tt.wantName)
					}
				}
			}

			for _, sub := range tt.wantTextContains {
				if !strings.Contains(got.Text, sub) {
					t.Errorf("Text missing %q, got: %s", sub, got.Text)
				}
			}

			if tt.wantMsgExact != "" {
				if got.Message != tt.wantMsgExact {
					t.Errorf("Message = %q, want %q", got.Message, tt.wantMsgExact)
				}
			}
			if tt.wantMsgContains != "" {
				if !strings.Contains(got.Message, tt.wantMsgContains) {
					t.Errorf("Message missing %q, got: %s", tt.wantMsgContains, got.Message)
				}
			}
		})
	}
}
